Revision:34660

Bacchusisi opened the pull request:
【bug】【CPP驱动】【游标】cpp驱动关闭所有游标时，发生迭代器越界的情况
【问题描述】 
cpp closeAllCursors接口的实现如下： 
  ```
 INT32 _sdbImpl::closeAllCursors() 
   { 
      INT32 rc = SDB_OK ; 

      // set all the cursors' status to be closed 
      for ( std::set<ossValuePtr>::iterator it = _cursors.begin(); 
            it != _cursors.end(); ++it ) 
      { 
         rc = ((_sdbCursorImpl *)(*it))->close() ; 
         if ( rc ) 
         { 
            goto error ; 
         } 
      } 
   done : 
      return rc ; 
   error : 
      goto done ; 
   } 
```

然而，` rc = ((_sdbCursorImpl *)(*it))->close() ;`会对_cursors中的指针元素进行删除，导致外层调用++it时会发生迭代器的越界的情况。

Bacchusisi commented:
【修改方法】 
将原方法修改如下： 
  ```
 INT32 _sdbImpl::closeAllCursors() 
   { 
      INT32 rc = SDB_OK ; 

      // set all the cursors' status to be closed 
      while( TRUE ) 
      { 
         std::set<ossValuePtr>::iterator it = _cursors.begin(); 
         if ( it != _cursors.end() ) 
         { 
            // for cursor.close() will unregister itself from "_cursors" 
            // so, we don't need to ++it to get next cursor 
            rc = ((_sdbCursorImpl *)(*it))->close() ; 
            if ( rc ) 
            { 
               goto error ; 
            } 
         } 
         else 
         { 
            break ; 
         } 
      } 
   done : 
      return rc ; 
   error : 
      goto done ; 
   } 
```

【提交】 
```
v2.8 
Command: Commit 
Modified: E:\sequoiadb\SequoiaDB\engine\client\clientcpp.cpp 
Sending content: E:\sequoiadb\SequoiaDB\engine\client\clientcpp.cpp 
Completed: At revision: 27083 
v2.6 
Command: Commit 
Modified: E:\sequoiadb2.6\SequoiaDB\engine\client\clientcpp.cpp 
Sending content: E:\sequoiadb2.6\SequoiaDB\engine\client\clientcpp.cpp 
Completed: At revision: 27084 
v1.12 
Command: Commit 
Modified: E:\sequoiadb1.12\SequoiaDB\engine\client\clientcpp.cpp 
Sending content: E:\sequoiadb1.12\SequoiaDB\engine\client\clientcpp.cpp 
Completed: At revision: 27085 
```

【测试】 
1、在windows下编译debug版的cpp动态库。 
2、使用sdbcppd.dll及sdbcppd.lib来编译运行sample中的connectWithSdbDataSource.cpp。 
   编译方式请参考connectWithSdbDataSource.cpp的文件头部的信息。 
3、如果顺利运行，说明修复正确。 
4、如果发生coredump，并且失败原因是因为访问set容器的迭代器越界，说明问题还没修复。 

【check list】 
检查项---------------------------------------------------------结论（NA/N/Y） 
修改代码是否全部提交SVN主干和分支版本 
且符合代码规范-------------------------------------------- 提交到1.12，2.6，2.8。 
是否用例全部通过----------------------------------------- Y 
是否清淅标明重现步骤---------------------------------- Y 
修改代码是否通过检视------------------------------------ N 
资料是否变更----------------------------------------------- NA 
接口是否变更----------------------------------------------- NA 
相关驱动是否变更或添加--------------------------------- NA 
是否保持向前兼容------------------------------------------ NA 
是否影响安装部署和监控--------------------------------- NA 
是否对影响性能-------------------------------------------- N 
sdb shell帮助信息是否更新------------------------------ NA

Bacchusisi commented:
请安排检视。

FlorentinoJH commented:
先拷贝一份临时set

Bacchusisi commented:
【提交】 
```
v2.8 
/trunk/SequoiaDB/engine/client/clientImpl.hpp 
/trunk/SequoiaDB/engine/client/clientcpp.cpp 
/trunk/SequoiaDB/engine/client/sdbDataSourceComm.hpp 
Completed: At revision: 27234 
v2.6 
Command: Commit 
Modified: E:\sequoiadb2.6\SequoiaDB\engine\client\clientcpp.cpp 
Sending content: E:\sequoiadb2.6\SequoiaDB\engine\client\clientcpp.cpp 
Completed: At revision: 27335 
v1.12 
Command: Commit 
Modified: E:\sequoiadb1.12\SequoiaDB\engine\client\clientcpp.cpp 
Sending content: E:\sequoiadb1.12\SequoiaDB\engine\client\clientcpp.cpp 
Completed: At revision: 27334 
```


【修改方法】 
1、v2.8版本包含连接池，所以，closeAllCursors方法通过发送interrupt消息来关闭context资源。 
2、v1.2和v2.6版本不包含连接池，closeAllCursors方法先拷贝一份std::set<ossValuePtr>，然后 
再对拷贝逐个遍历，逐个释放游标资源。 


【测试】 
1、在windows下编译debug版的cpp动态库。 
2、使用sdbcppd.dll及sdbcppd.lib来编译运行sample中的connectWithSdbDataSource.cpp。 
   编译方式请参考connectWithSdbDataSource.cpp的文件头部的信息。 
3、如果顺利运行，说明修复正确。 
4、如果发生coredump，并且失败原因是因为访问set容器的迭代器越界，说明问题还没修复。 

【check list】 
检查项---------------------------------------------------------结论（NA/N/Y） 
修改代码是否全部提交SVN主干和分支版本 
且符合代码规范-------------------------------------------- 提交到1.12，2.6，2.8。 
是否用例全部通过----------------------------------------- Y 
是否清淅标明重现步骤---------------------------------- Y 
修改代码是否通过检视------------------------------------ Y 
资料是否变更----------------------------------------------- NA 
接口是否变更----------------------------------------------- NA 
相关驱动是否变更或添加--------------------------------- NA 
是否保持向前兼容------------------------------------------ NA 
是否影响安装部署和监控--------------------------------- NA 
是否对影响性能-------------------------------------------- N 
sdb shell帮助信息是否更新------------------------------ NA

Bacchusisi commented:
请转测试。

FlorentinoJH commented:
请安排测试

EmilianLLJH commented:
请测试

BrigidLLX commented:
【验证结果】：问题未解决 
【验证步骤】 
1、分别创建普通连接、连接池分配的连接 
2、在连接上创建cs，cl 
3、对cl做插入查询操作获得cursor 
4、对于普通连接，调用closeAllCursors关闭游标；对于连接池分配的连接，通过调用releaseConnection的方式间接调用closeAllCursors关闭游标 
5、调用next方法检查cursor失效 
6、删除cs,cl，断开连接，关闭连接池等。 
具体测试用例参见附件closeAllCursors.cpp，在windows上执行该用例时发生core dump。 
检查发现，在用例中调用closeAllCursors会将cursor中的_collection置为NULL，当用例结束退出执行cursor的析构方法时导致cursor与cl不能解除关联，此时cl关联的cursor为一个无效对象。当执行到cl的析构函数时，会产生core dump。 
详细信息请参考附件。

EmilianLLJH commented:
问题还未解决

FlorentinoJH commented:
请全面处理。

Bacchusisi commented:
【问题原因】 
关闭所有游标回话后，需要在db/cl上的做游标的反注册工作。在做这个反注册工作过程，出现如下两个问题： 
1、在对set遍历的过程，又对set进行了删除，导致set的状态紊乱，导致coredump。 
2、游标在close过程或者在析构过程，存在反注册失败的情况，导致db/cl析构时，出现访问非法内存的情况。 

【修改】 
正对上述两个问题，修改如下： 
问题1： 
对set容器修改，先做临时拷贝，然后再做遍历，最后再将原set容器内容清空。 
问题2： 
确保游标（或者lob）在close()方法中，或者在析构方法中，必然能够成功进行反注册。 

【提交】 
```
v2.9 
Command: Commit 
Modified: E:\sequoiadb\SequoiaDB\engine\client\clientImpl.hpp 
Modified: E:\sequoiadb\SequoiaDB\engine\client\clientcpp.cpp 
Sending content: E:\sequoiadb\SequoiaDB\engine\client\clientcpp.cpp 
Sending content: E:\sequoiadb\SequoiaDB\engine\client\clientImpl.hpp 

v2.8 
Command: Commit 
Modified: F:\sequoiadb_2.8\SequoiaDB\engine\client\clientImpl.hpp 
Modified: F:\sequoiadb_2.8\SequoiaDB\engine\client\clientcpp.cpp 
Sending content: F:\sequoiadb_2.8\SequoiaDB\engine\client\clientcpp.cpp 
Sending content: F:\sequoiadb_2.8\SequoiaDB\engine\client\clientImpl.hpp 

v2.6 
Command: Commit 
Modified: E:\sequoiadb2.6\SequoiaDB\engine\client\clientImpl.hpp 
Modified: E:\sequoiadb2.6\SequoiaDB\engine\client\clientcpp.cpp 
Sending content: E:\sequoiadb2.6\SequoiaDB\engine\client\clientcpp.cpp 
Sending content: E:\sequoiadb2.6\SequoiaDB\engine\client\clientImpl.hpp 
```

【check list】 
检查项结论（NA/N/Y）
修改代码是否全部提交SVN主干和分支版本且符合代码规范--- v2.9，v2.8，v2.6 
是否用例全部通过-------------------------------------- Y 
是否清淅标明重现步骤---------------------------------- NA 
修改代码是否通过检视---------------------------------- N 
资料是否变更------------------------------------------ Y 
接口是否变更------------------------------------------ Y 
相关驱动是否变更或添加-------------------------------- NA 
是否保持向前兼容-------------------------------------- Y 
是否影响安装部署和监控-------------------------------- NA 
是否对影响性能---------------------------------------- NA 
sdb shell帮助信息是否更新----------------------------- NA

Bacchusisi commented:
请安排检视和测试。测试过程，如有疑问，请及时沟通。

Bacchusisi commented:
`v2.6.4 closeAllCursor（）`接口并没有考虑lob的情况。该问题已在3326描述。故不在此单处理。

FlorentinoJH commented:
请安排测试

EmilianLLJH commented:
请验证

BrigidLLX commented:
【验证结果】：验证问题已经解决 
【验证步骤】 
1、在windows下编译debug版本的C++驱动动态库 
2、编译执行samples下的connectWithSdbDataSource.cpp，结果正常 
3、编译执行测试用例closeAllCursors.cpp，结果正常 
4、在linux下执行上述步骤，结果相同 
【测试用例】 
TestLink： 
14813:执行closeAllCursor关闭游标 
```
SVN： 
trunk\testcase_new\driver\cpp\cursor\closeAllCursors14183.cpp 
branches\sequoiadb_2.8\testcase_new\driver\cpp\cursor\closeAllCursors14183.cpp 
branches\sequoiadb_2.6\testcases\hlt\drivers_testcases\cpp\closeAllCursors.cpp
```

BrigidLLX commented:
请检视

AydenYYD commented:
已经确认，之前是必现问题，相同的验证方式验证没再出现。
